pkg/core/basic_chain_test.go

```diff
@@ -596,3 +625,1634 @@ func TestBlockchain_StartFromExistingDB(t *testing.T) {
+	require.True(t, len(notificationCh) >= 4)
+	for i := 0; i < 4; i++ {
+		notif := <-notificationCh
+		require.Equal(t, nativeGASHash, notif.ScriptHash)
+	}
+
+	// Follow in-block transaction order.
+	for _, txExpected := range invBlock.Transactions {
+		tx := <-txCh
+		require.Equal(t, txExpected, tx)
+		exec := <-executionCh
+		require.Equal(t, tx.Hash(), exec.Container)
+		if exec.VMState == vm.HaltState {
+			notif := <-notificationCh
+			require.Equal(t, hash.Hash160(tx.Script), notif.ScriptHash)
+		}
+	}
+	assert.Empty(t, txCh)
+	assert.Len(t, notificationCh, 1)
+	assert.Len(t, executionCh, 1)
+
+	notif = <-notificationCh
+	require.Equal(t, bc.UtilityTokenHash(), notif.ScriptHash)
+
+	exec = <-executionCh
+	require.Equal(t, b.Hash(), exec.Container)
+	require.Equal(t, exec.VMState, vm.HaltState)
+
+	bc.UnsubscribeFromBlocks(blockCh)
+	bc.UnsubscribeFromTransactions(txCh)
+	bc.UnsubscribeFromNotifications(notificationCh)
+	bc.UnsubscribeFromExecutions(executionCh)
+
+	// Ensure that new blocks are processed correctly after unsubscription.
+	e.GenerateNewBlocks(t, 2*chBufSize)
+}
+
+func TestBlockchain_RemoveUntraceable(t *testing.T) {
+	check := func(t *testing.T, bc *core.Blockchain, tHash, bHash, sHash util.Uint256, errorExpected bool) {
+		_, _, err := bc.GetTransaction(tHash)
+		if errorExpected {
+			require.Error(t, err)
+		} else {
+			require.NoError(t, err)
+		}
+		_, err = bc.GetAppExecResults(tHash, trigger.Application)
+		if errorExpected {
+			require.Error(t, err)
+		} else {
+			require.NoError(t, err)
+		}
+		_, err = bc.GetBlock(bHash)
+		if errorExpected {
+			require.Error(t, err)
+		} else {
+			require.NoError(t, err)
+		}
+		_, err = bc.GetHeader(bHash)
+		require.NoError(t, err)
+		if !sHash.Equals(util.Uint256{}) {
+			sm := bc.GetStateModule()
+			_, err = sm.GetState(sHash, []byte{0xfb, 0xff, 0xff, 0xff, 0x0e}) // NEO committee key.
+			if errorExpected {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+		}
+	}
+	t.Run("P2PStateExchangeExtensions off", func(t *testing.T) {
+		bc, acc := chain.NewSingleWithCustomConfig(t, func(c *config.ProtocolConfiguration) {
+			c.MaxTraceableBlocks = 2
+			c.GarbageCollectionPeriod = 2
+			c.RemoveUntraceableBlocks = true
+		})
+		e := neotest.NewExecutor(t, bc, acc, acc)
+		neoValidatorInvoker := e.ValidatorInvoker(e.NativeHash(t, nativenames.Neo))
+
+		tx1Hash := neoValidatorInvoker.Invoke(t, true, "transfer", acc.ScriptHash(), util.Uint160{1, 2, 3}, 1, nil)
+		tx1Height := bc.BlockHeight()
+		b1 := e.TopBlock(t)
+		sRoot, err := bc.GetStateModule().GetStateRoot(tx1Height)
+		require.NoError(t, err)
+
+		neoValidatorInvoker.Invoke(t, true, "transfer", acc.ScriptHash(), util.Uint160{1, 2, 3}, 1, nil)
+
+		_, h1, err := bc.GetTransaction(tx1Hash)
+		require.NoError(t, err)
+		require.Equal(t, tx1Height, h1)
+
+		check(t, bc, tx1Hash, b1.Hash(), sRoot.Root, false)
+		e.GenerateNewBlocks(t, 4)
+
+		// Wait for Run().
+		time.Sleep(2 * bcPersistInterval)
```

Comment:
Unacceptable. You can `require.Eventually()`, but even that is far from perfect, it just takes time.